osmoutils/partialord/internal/dag/dag.go

```diff
@@ -0,0 +1,305 @@
+	dag.directedEdgeList[v][u] = -1
+	return nil
+}
+
+// replaceEdge adds a directed edge from u -> v.
+// it removes any edge that may already exist between the two.
+func (dag *Dag) replaceEdge(u, v int) error {
+	if u == v {
+		return fmt.Errorf("can't make self-edge")
+	}
+
+	dag.directedEdgeList[u][v] = 1
+	dag.directedEdgeList[v][u] = -1
+	return nil
+}
+
+// deleteEdge deletes edges between u and v.
+func (dag *Dag) deleteEdge(u, v int) {
+	delete(dag.directedEdgeList[u], v)
+	delete(dag.directedEdgeList[v], u)
+}
+
+func (dag *Dag) AddEdge(u, v string) error {
+	uIndex, uExists := dag.nodeNames[u]
+	vIndex, vExists := dag.nodeNames[v]
+	if !uExists || !vExists {
+		return fmt.Errorf("one of %s, %s does not exist in dag", u, v)
+	}
+	return dag.addEdge(uIndex, vIndex)
+}
+
+// ReplaceEdge adds a directed edge from u -> v.
+// it removes any edge that may already exist between the two.
+func (dag *Dag) ReplaceEdge(u, v string) error {
+	uIndex, uExists := dag.nodeNames[u]
+	vIndex, vExists := dag.nodeNames[v]
+	if !uExists || !vExists {
+		return fmt.Errorf("one of %s, %s does not exist in dag", u, v)
+	}
+	return dag.replaceEdge(uIndex, vIndex)
+}
+
+// AddFirstElements sets the provided elements to be first in all orderings.
+// So if were making an ordering over {A, B, C, D, E}, and elems provided is {D, B, A}
+// then we are guaranteed that the total ordering will begin with {D, B, A}
+func (dag *Dag) AddFirstElements(nodes ...string) error {
+	nodeIds, err := dag.namesToIds(nodes)
+	if err != nil {
+		return err
+	}
+
+	return dag.addFirst(nodeIds)
+}
+
+func (dag *Dag) addFirst(nodes []int) error {
+	nodeMap := map[int]bool{}
+	for i := 0; i < len(nodes); i++ {
+		nodeMap[nodes[i]] = true
+	}
+	// First we add an edge from nodes[-1] to every node in the graph aside from the provided first nodes.
+	// then we make nodes[-1] depend on nodes[-2], etc.
+	lastOfFirstNodes := nodes[len(nodes)-1]
+	for i := 0; i < len(dag.nodeNames); i++ {
+		// skip any node in the 'first set'
+		_, inMap := nodeMap[i]
+		if inMap {
+			continue
+		}
+		// We make everything on u, and therefore have an edge from u to it
+		err := dag.replaceEdge(lastOfFirstNodes, i)
+		// can't happen by above check
+		if err != nil {
+			return err
+		}
+	}
+
+	// Make nodes[i+1] depend on nodes[i]
+	for i := len(nodes) - 2; i >= 0; i-- {
+		err := dag.replaceEdge(nodes[i], nodes[i+1])
+		// can't happen by above check
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddLastElements sets the provided elements to be last in all orderings.
+// So if were making an ordering over {A, B, C, D, E}, and elems provided is {D, B, A}
+// then we are guaranteed that the total ordering will end with {D, B, A}
+func (dag *Dag) AddLastElements(nodes ...string) error {
+	nodeIds, err := dag.namesToIds(nodes)
+	if err != nil {
+		return err
+	}
+
+	return dag.addLast(nodeIds)
+}
+
+func (dag *Dag) addLast(nodes []int) error {
+	nodeMap := map[int]bool{}
```

Comment:
nit:
```suggestion
	nodeMap := map[int]struct{}
```

Reply:
Is this more idiomatic for existence?

Reply:
Since there are no sets in Go, this idiom is commonly used in place of sets where it avoids spending an unnecessary byte for `bool` in each key/value pair:
- https://www.davidkaya.com/sets-in-golang/
- https://dave.cheney.net/2014/03/25/the-empty-struct

Reply:
I don't think this is a nit that matters. (Nor do I think either article leads one to the conclusion around whats idiomatic in golang. The Dave Cheney blog doesn't advocate for this in a map. Go Nuts developer threads don't have clear consensus either.

In a larger scheme, this is a micro-optimization, that has no impact on system performance. Do you find this a readability hurdle?

Reply:
I generally try to avoid leaving out micro-optimizations if there is a language construct that allows mitigating the overhead. As the repository evolves, the use patterns change and the overhead compounds, such things might become problematic.

I'm also used to seeing this pattern across codebases so it doesn't feel unreadable to me. 

However, since all my points are a matter of personal preference and trying to optimize something negligible, I'm okay with keeping everything as is.

Thanks for being thorough in addressing and discussing this comment :+1: